File: unique-length-3-palindromic-subsequences/unique-length-3-palindromic-subsequences.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        
        unordered_set<char> p;
        
        int first[26], last[26], count = 0;
        
        fill_n(first, 26, INT_MAX);
        fill_n(last, 26, 0);
        
        for(int i = 0; i < s.size(); ++i) {
            
            int letter = s[i] - 'a';
            
            first[letter] = min(first[letter], i);
            last[letter] = i;
        }
        
        for (int j = 0; j < 26; ++j) {
            
            if (first[j] < INT_MAX) {
            
                p.clear();
                
                for (int k = first[j] + 1; k < last[j]; ++k) { p.insert(s[k]); }

                count += p.size();
            }
        }
        
        return count;
    }
};

File: unique-length-3-palindromic-subsequences/unique-length-3-palindromic-subsequences.java
class Solution {
    public int countPalindromicSubsequence(String s) {
        
        int count = 0;
        
        int first[] = new int[26], last[] = new int[26]; 
        
        Arrays.fill(first, Integer.MAX_VALUE);
        
        
        for(int i = 0; i < s.length(); ++i) {
            
            int letter = s.charAt(i) - 'a';
            
            first[letter] = Math.min(first[letter], i);
            last[letter] = i;
        }
        
        
        for (int j = 0; j < 26; ++j) {
            
            if (first[j] < last[j]) {

                count += s.substring(first[j] + 1, last[j]).chars().distinct().count();
            }
        }
        
        
        return count;
    }
}